DAGE-4: Added code to deal with config file

nicolo-rinaldi opened the pull request:
## Notes
Added code to deal with config file (parsing, testing, CLI interface). Code added:
- config.py
- test_config.py
- all files needed to test an initial config
- added logging to 'config.py' and 'dataset_generator.py' files

## Ticket
https://sease.atlassian.net/jira/software/projects/DAGE/boards/84?selectedIssue=DAGE-4

nicolo-rinaldi committed:
[DAGE-4] Configuration file and validation of config params + test

## Changes
- created the 'config.py' script with the Config class used to deal with the input config file. pydantic has been used, that already includes type validation
- created file 'test_config.py' in the folder 'test/unit/' to test the code for the config file handling/parsing
- added all the file needed for the 'test_config.py' in the folder 'test/unit/resources/' and in the main folder
- updated 'pyproject.toml' file with the installation of the 'argparse' library
- updated main file ('dataset_generator.py') to take the config.yaml file as a CLI input
- updated the 'README.md' file with an easier command to execute to run tests

nicolo-rinaldi committed:
[DAGE-4] Configuration file and validation of config params + test (update)

## Changes
- added logging to the files 'config.py' and 'dataset_generator.py'

nseidan commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+# Must be at least one field
+docFields:
+  - "title"
+  - "body"
```
we have "description" field instead of "body"

nseidan commented on rre-dataset-generator/dataset_generator.py:
```diff
@@ -1,29 +1,27 @@
-        log.debug(f"Dummy variable {args.dummy} loaded successfully.")
+        config = load_config(args.config_file)
+        log.debug("Configuration loaded successfully.")
+        # print(config.model_dump_json(indent=2))
```
redundant # print(config.model_dump_json(indent=2)) ?

nseidan commented on rre-dataset-generator/src/config.py:
```diff
@@ -0,0 +1,71 @@
+
+
+class Config(BaseModel):
+    QueryTemplate: Optional[str] = Field("q=#$query##", description="Template string for queries with a placeholder for keywords.")
```
naming convention: better to use  lowercase with underscores for field names e.g. query_template, search_engine_type ("solr", "vespa" ...) ?

dantuzi replied:
+1
I would follow this standard: https://google.github.io/styleguide/pyguide.html#316-naming

nseidan commented on rre-dataset-generator/src/config.py:
```diff
@@ -0,0 +1,71 @@
+    @field_validator('LLMConfigurationFile')
+    def check_config_type(cls, v):
+        if v is not None:
+            if v.suffix[1:] != "yaml" and v.suffix[1:] != "yml":
```
more readable:  `if v.suffix[1:] not in {"yaml", "yml"}:`

nicolo-rinaldi committed:
DAGE-4: updated naming and deleted typos

## Changes
- changed naming convention in 'config.py' to be more python-like
- updated all the files in 'resources/' + 'config.yaml' to be compliant with new naming convention
- changed 'test_config.py' file to be compliant with new naming convention

dantuzi commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+# (Optional) Template string for generated/given queries used to search documents and build relevance
+# judgemenets, with a placeholder for inserting keywords
+# Example (Solr):
```
This is not the query used to search documents and build relevance judgements.

This is the query the user wants to evaluate, and to do that, we are generating the entire dataset.

nicolo-rinaldi replied:
Changed description. I did not understood fully the use of this param. Now should be fine

dantuzi commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+
+# The type of search engine to use
+# Accepted values: Solr, Elasticsearch, Opensearch, Vespa
+search_engine_type: "solr"
```
Are the accepted values uppercase or lowercase?
The documentation says we accept "Solr" but the example value is "solr".
Is it case-insensitive?

nicolo-rinaldi replied:
I forgot to change it. I did it now

nseidan commented on rre-dataset-generator/src/config.py:
```diff
@@ -5,25 +5,25 @@
+    total_num_queries_to_generate: int = Field(..., gt=0, description="Total number of queries to generate.")
+    relevance_scale: Literal['binary', 'graded']
+    llm_configuration_file: FilePath = Field(..., description="Path to the LLM configuration file.")
+    output_format: Literal['Quepid', 'RRE']
```
@nicolo-rinaldi minor, to lowercase -> quepid, rre ? to keep it consistent with others like binary/graded etc

nseidan commented on rre-dataset-generator/tests/unit/resources/invalid_type.yaml:
```diff
@@ -1,9 +1,9 @@
+relevance_scale: "binary"
+llm_configuration_file: "llm_config.yaml"
+output_format: "RRE"
+output_destination: "out.json"
```
@nicolo-rinaldi rename to output.json ? more meaningful.

dantuzi commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+search_engine_collection_endpoint: "http://localhost:8983/solr/mycore"
+
+# (Optional) Filter query to restrict the set of documents used to generate queries. Top k retrieved documents
+# are used, with k computed accordingly looking at the params: docNumber, totalNumQueriesToGenerate
```
When we are retrieving documents, we only use `docNumber`.
Maybe I understand why you also add the parameter `totalNumQueriesToGenerate`, you want to say that if the number of provided queries is higher than the `totalNumQueriesToGenerate`, we don't need to extract any document from Solr. In that case, I would avoid this at this stage because it can be confusing.
I wouldn't say anything about getting the "top k" documents, but if you want to say that, I would only mention the docNumber.

nicolo-rinaldi replied:
removed the mention to top k retrieved docs

nicolo-rinaldi committed:
[DAGE-4] Minor changes to test files + output_format

dantuzi commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+generate_queries_from_documents: true
+
+# Total number of queries to generate (includes predefined queries, if any)
+total_num_queries_to_generate: 10
```
I'm sorry for mentioning it now, but I believe this name can be misleading.
If I read "total number of queries to generate" I'm expecting the software always generates that number of queries (but it can be less if there are enough provided queries).
How about something more generic like `num_queries` or `num_queries_needed` (any better idea is welcome)

nicolo-rinaldi replied:
I changed it in `num_queries_needed`

dantuzi commented on rre-dataset-generator/config.yaml:
```diff
@@ -0,0 +1,64 @@
+
+# Whether to generate an additional file with query, document, score, and explanation
+# Accepted values: true, false
+output_explanation: true
```
I remember we discussed having two parameters instead of just one to be added for version V2. [Notes](https://docs.google.com/document/d/1oCofhKdphXUJgQJOvLZ-NLeVvirLMm84JATXJR8jAOI/edit?tab=t.0)
```
Add parameters for reasoning (for the LLM) and save_explanation (save the explanation file) if save_explanation=true => reasoning=true
```

That being said, I'd remove this parameter for now, and we can add it as part of the V2 implementation task

nicolo-rinaldi replied:
Ok, an idea can be also to put this param inside the `llm_config` file. I don't know if it make sense since is an option that reflects also in the structure of the output dataset

nicolo-rinaldi replied:
Anyway, for now i removed the `output_explanation` param

dantuzi commented on rre-dataset-generator/llm_config.yaml:
Why do we need to commit this empty file?

nicolo-rinaldi replied:
I used it to run test, since pydantic datatype does automatic validation and needs an actual file. Maybe I can move it to the `tests/unit/resources/`?

nicolo-rinaldi replied:
I decided to move the file in the mentioned folder

dantuzi commented on rre-dataset-generator/queries.txt:
Same as before. We shouldn't commit this file because the user can link any file

nicolo-rinaldi replied:
Same as last comment

dantuzi commented on rre-dataset-generator/pyproject.toml:
```diff
@@ -5,6 +5,7 @@ description = "Dataset Generator project"
 readme = "README.md"
 requires-python = ">=3.10"
 dependencies = [
+    "argparse>=1.4.0",
```
Did we forget to add this in the previous PR?

nicolo-rinaldi replied:
It should be in python base library, but it was raising me some errors, so I updated the library `argparse`

dantuzi replied:
Just out of curiosity, what error did you get?

dantuzi commented on rre-dataset-generator/src/config.py:
```diff
@@ -0,0 +1,71 @@
+        raw_config = yaml.safe_load(f)
+    return Config(**raw_config)
+    # except Exception as e:
+    #     raise RuntimeError(f"Error loading config: {e}")
```
Commented lines like these are not needed.
I think an exception with a descriptive message may be useful. Is it already provided by `safe_load` function?

nicolo-rinaldi replied:
Yes, I commented this code to have more descriptive exception. I removed now also the comments

dantuzi commented on rre-dataset-generator/src/config.py:
```diff
@@ -0,0 +1,71 @@
+    #     raise RuntimeError(f"Error loading config: {e}")
+
+
+# Example usage in pipeline
+if __name__ == "__main__":
+    from src.logger import configure_logging
+    import logging
+
+    configure_logging(level=logging.DEBUG)
+    log = logging.getLogger(__name__)
+
+    config = load_config("config.yaml")
+    log.debug("Configuration loaded successfully.")
```
Why this? Sorry for my ignorance, but isn't it possible to test this using the unit tests?

nicolo-rinaldi replied:
Yes, it is. I removed this code

dantuzi commented on rre-dataset-generator/tests/unit/test_config.py:
```diff
@@ -0,0 +1,82 @@
+    assert config.output_destination == Path("output/generated_dataset.json")
+
+def test_output_explanation(config):
+    assert config.output_explanation is True
```
All of those tests can be written in a single test case.

```
def test_good_config_expect_all_parameters_read:
    config = load_config(good file)
    assert parameter 1
    assert parameter 2
    ...
    assert parameter n
```

nicolo-rinaldi replied:
Done

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -13,4 +11,4 @@ def args():
 def test_argparse_with_defaults(args):
-    assert args.dummy == "default"
\ No newline at end of file
+    assert args.config_file == "config.yaml"
```
Can you please add a new line at the end?

nicolo-rinaldi replied:
done

nicolo-rinaldi committed:
[DAGE-4] Modification on PR https://github.com/SeaseLtd/rated-ranking-evaluator/pull/152

## Changes
- moved empty files used for tests to the 'tests/unit/resources/' folder
- modified tests to be more compact and readable
- removed 'output_explaination' from the params
- other minor changes

dantuzi commented on rre-dataset-generator/tests/unit/test_dataset_generation_dummy.py:
```diff
@@ -1,5 +1,3 @@
-import sys
-import os
 from src.logger import configure_logging
```
it is used
